### ISSUE TITLE
Illegal memory access when training HuggingFace model on GPU with Inductor

### ISSUE THREAD

**Issue by Lokiiiiii** [pytorch/pytorch#95750]
### 🐛 Describe the bug

When using HuggingFace's Trainer API I noticed that PyTorch eager mode succeeds as expected but inductor fails with an illegal memory access.

### Error logs

| 1677525140847 | Traceback (most recent call last): File "./run_mlm.py", line 694, in |
| 1677525140847 | main() File "./run_mlm.py", line 635, in main |
| 1677525140847 | train_result = trainer.train(resume_from_checkpoint=checkpoint) File "/opt/conda/lib/python3.8/site-packages/transformers/trainer.py", line 1631, in train |
| 1677525140847 | return inner_training_loop( File "/opt/conda/lib/python3.8/site-packages/transformers/trainer.py", line 1898, in _inner_training_loop |
| 1677525140848 | tr_loss_step = self.training_step(model, inputs) File "/opt/conda/lib/python3.8/site-packages/transformers/trainer.py", line 2640, in training_step |
| 1677525140848 | loss = self.compute_loss(model, inputs) File "/opt/conda/lib/python3.8/site-packages/transformers/trainer.py", line 2672, in compute_loss |
| 1677525140848 | outputs = model(**inputs) File "/opt/conda/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 82, in call return self.dynamo_ctx(self._orig_mod.call)(*args, **kwargs) File "/opt/conda/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 215, in _fn |
| 1677525140848 | return fn(*args, **kwargs) File "/opt/conda/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl |
| 1677525140848 | return forward_call(*args, **kwargs) File "/opt/conda/lib/python3.8/site-packages/transformers/models/bert/modeling_bert.py", line 1324, in forward |
| 1677525140848 | @add_start_docstrings_to_model_forward(BERT_INPUTS_DOCSTRING.format("batch_size, sequence_length")) File "/opt/conda/lib/python3.8/site-packages/torch/_dynamo/eval_frame.py", line 215, in _fn |
| 1677525140848 | return fn(*args, **kwargs) File "/opt/conda/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2816, in forward |
| 1677525140848 | return compiled_fn(full_args) File "/opt/conda/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 1222, in g |
| 1677525140848 | return f(*args) File "/opt/conda/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2383, in debug_compiled_function |
| 1677525140848 | return compiled_function(*args) File "/opt/conda/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 1895, in runtime_wrapper |
| 1677525140848 | all_outs = call_func_with_args( File "/opt/conda/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 1256, in call_func_with_args |
| 1677525140848 | out = normalize_as_list(f(*args)) File "/opt/conda/lib/python3.8/site-packages/torch/autograd/function.py", line 506, in apply |
| 1677525140848 | return super().apply(*args, **kwargs) # type: ignore[misc] File "/opt/conda/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 2148, in forward |
| 1677525140848 | fw_outs = call_func_with_args( File "/opt/conda/lib/python3.8/site-packages/torch/_functorch/aot_autograd.py", line 1247, in call_func_with_args |
| 1677525140848 | out = normalize_as_list(f(args)) File "/opt/conda/lib/python3.8/site-packages/torch/_inductor/compile_fx.py", line 248, in run |
| 1677525140848 | return model(new_inputs) File "/tmp/torchinductor_root/rj/crjch2m3bp6tuhd3s6n2apgbibxay4o6o5jlrfbwsfiokrv2rkep.py", line 4483, in call |
| 1677525140848 | triton__49.run(primals_208, buf513, buf514, 128, 128, grid=grid(128), stream=stream0) File "/opt/conda/lib/python3.8/site-packages/torch/_inductor/triton_ops/autotune.py", line 184, in run |
| 1677525140848 | self.autotune_to_one_config(*args, grid=grid) File "/opt/conda/lib/python3.8/site-packages/torch/_dynamo/utils.py", line 163, in time_wrapper |
| 1677525140848 | r = func(*args, **kwargs) File "/opt/conda/lib/python3.8/site-packages/torch/_inductor/triton_ops/autotune.py", line 171, in autotune_to_one_config timings = { File "/opt/conda/lib/python3.8/site-packages/torch/_inductor/triton_ops/autotune.py", line 172, in |
| 1677525140848 | launcher: self.bench(launcher, *cloned_args, **kwargs) File "/opt/conda/lib/python3.8/site-packages/torch/_inductor/triton_ops/autotune.py", line 153, in bench return do_bench(kernel_call, rep=40, fast_flush=True) File "/opt/conda/lib/python3.8/site-packages/triton/testing.py", line 144, in do_bench |
| 1677525140848 | torch.cuda.synchronize() File "/opt/conda/lib/python3.8/site-packages/torch/cuda/init.py", line 711, in synchronize |
| 1677525140848 | return torch._C._cuda_synchronize() |
| 1677525140848 | RuntimeError: CUDA error: an illegal memory access was encountered

### Minified repro

Minifier does not repro the error. 

TORCHDYNAMO_REPRO_AFTER=dynamo leads to minifier not finding an issue
TORCHDYNAMO_REPRO_AFTER=aot leads to a completely different error signature

### Repro

pip3 install numpy --pre torch --force-reinstall --index-url https://download.pytorch.org/whl/nightly/cu117
git clone https://github.com/huggingface/transformers.git
cd transformers
pip install -e .

cd examples/pytorch/language-modelling
pip install -r requirements.txt
WANDB_DISABLED=true python run_mlm.py --model_name_or_path bert-base-uncased --output_dir . --fp16 --dataloader_drop_last --dataset_config_name wikitext-2-raw-v1 --dataset_name wikitext --do_train --evaluation_strategy no --logging_strategy epoch --max_seq_length 128 --num_train_epochs 1 --overwrite_output_dir --per_device_train_batch_size 128 --save_strategy no --torch_compile_backend inductor

### Versions

PyTorch version: 2.0.0a0+git8693604
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.25.2
Libc version: glibc-2.31

Python version: 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 16:01:55)  [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-1028-aws-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA A10G
GPU 1: NVIDIA A10G
GPU 2: NVIDIA A10G
GPU 3: NVIDIA A10G
GPU 4: NVIDIA A10G
GPU 5: NVIDIA A10G
GPU 6: NVIDIA A10G

Nvidia driver version: 515.65.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          192
On-line CPU(s) list:             0-191
Thread(s) per core:              2
Core(s) per socket:              48
Socket(s):                       2
NUMA node(s):                    2
Vendor ID:                       AuthenticAMD
CPU family:                      23
Model:                           49
Model name:                      AMD EPYC 7R32
Stepping:                        0
CPU MHz:                         2799.534
BogoMIPS:                        5599.06
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       3 MiB
L1i cache:                       3 MiB
L2 cache:                        48 MiB
L3 cache:                        384 MiB
NUMA node0 CPU(s):               0-47,96-143
NUMA node1 CPU(s):               48-95,144-191
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy cr8_legacy abm sse4a misalignsse 3dnowprefetch topoext perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr rdpru wbnoinvd arat npt nrip_save rdpid

Versions of relevant libraries:
[pip3] clip-anytorch==2.5.0
[pip3] CoCa-pytorch==0.0.7
[pip3] dalle2-pytorch==1.10.5
[pip3] ema-pytorch==0.2.1
[pip3] functorch==1.14.0a0+408bcf1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.23.5
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-warmup==0.1.1
[pip3] rotary-embedding-torch==0.2.1
[pip3] sagemaker-pytorch-training==2.7.0
[pip3] torch==2.0.0a0+git8693604
[pip3] torch-fidelity==0.3.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.0.0a0+fd77809
[pip3] torchdata==0.5.1+a246b31
[pip3] torchmetrics==0.11.1
[pip3] torchrec-nightly==2023.2.24
[pip3] torchtext==0.14.0a0+5b78d07
[pip3] torchvision==0.15.0a0+31a4ef9
[pip3] vector-quantize-pytorch==1.0.4
[pip3] vit-pytorch==1.0.2
[conda] clip-anytorch             2.5.0                    pypi_0    pypi
[conda] coca-pytorch              0.0.7                    pypi_0    pypi
[conda] dalle2-pytorch            1.10.5                   pypi_0    pypi
[conda] ema-pytorch               0.2.1                    pypi_0    pypi
[conda] functorch                 1.14.0a0+408bcf1          pypi_0    pypi
[conda] magma-cuda117             2.6.1                         1    pytorch
[conda] mkl                       2022.2.1         h84fe81f_16997    conda-forge
[conda] mkl-include               2023.0.0         h84fe81f_26648    conda-forge
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch                   1.13.1          cpu_py38hbac4b8a_1    conda-forge
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-warmup            0.1.1                    pypi_0    pypi
[conda] rotary-embedding-torch    0.2.1                    pypi_0    pypi
[conda] sagemaker-pytorch-training 2.7.0                    pypi_0    pypi
[conda] torch                     2.0.0a0+git8693604          pypi_0    pypi
[conda] torch-fidelity            0.3.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                2.0.0a0+fd77809          pypi_0    pypi
[conda] torchdata                 0.5.1            py38h60d003c_1    conda-forge
[conda] torchmetrics              0.11.1                   pypi_0    pypi
[conda] torchrec-nightly          2023.2.24                pypi_0    pypi
[conda] torchtext                 0.14.0a0+5b78d07          pypi_0    pypi
[conda] torchvision               0.15.0a0+31a4ef9          pypi_0    pypi
[conda] vector-quantize-pytorch   1.0.4                    pypi_0    pypi
[conda] vit-pytorch               1.0.2                    pypi_0    pypi

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

**Comment by mlazos** [comment]
I wasn't able to repro this error, but I encountered another on the nightly build while trying to repro this one. 

Repro steps: 
Follow the steps provided above, and also the step [here](https://github.com/microsoft/DeepSpeed/issues/2845) (remove references to torch._six) on pytorch commit (fafb410985d2cb94bd95f12f0c392bad9385b643)

[output](https://gist.github.com/mlazos/46886b75d6a3f1e6534c76f1e852ecd2)

The exception is this:
```
  File "/scratch/mlazos/pytorch/torch/_dynamo/output_graph.py", line 708, in call_user_compiler
    compiled_fn = compiler_fn(gm, self.fake_example_inputs())
  File "/scratch/mlazos/pytorch/torch/_dynamo/debug_utils.py", line 1055, in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
  File "/scratch/mlazos/pytorch/torch/__init__.py", line 1393, in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
  File "/scratch/mlazos/pytorch/torch/_inductor/compile_fx.py", line 488, in compile_fx
    return aot_autograd(
  File "/scratch/mlazos/pytorch/torch/_dynamo/backends/common.py", line 48, in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
  File "/scratch/mlazos/pytorch/torch/_functorch/aot_autograd.py", line 2807, in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
  File "/scratch/mlazos/pytorch/torch/_dynamo/utils.py", line 164, in time_wrapper
    r = func(*args, **kwargs)
  File "/scratch/mlazos/pytorch/torch/_functorch/aot_autograd.py", line 2500, in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config)
  File "/scratch/mlazos/pytorch/torch/_functorch/aot_autograd.py", line 1674, in aot_wrapper_dedupe
    fw_metadata, _out = run_functionalized_fw_and_collect_metadata(
  File "/scratch/mlazos/pytorch/torch/_functorch/aot_autograd.py", line 606, in inner
    flat_f_outs = f(*flat_f_args)
  File "/scratch/mlazos/pytorch/torch/_functorch/aot_autograd.py", line 2778, in functional_call
    out = Interpreter(mod).run(*args[params_len:], **kwargs)
  File "/scratch/mlazos/pytorch/torch/fx/interpreter.py", line 136, in run
    self.env[node] = self.run_node(node)
  File "/scratch/mlazos/pytorch/torch/fx/interpreter.py", line 177, in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
  File "/scratch/mlazos/pytorch/torch/fx/interpreter.py", line 294, in call_module
    return submod(*args, **kwargs)
  File "/scratch/mlazos/pytorch/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/scratch/mlazos/pytorch/torch/nn/modules/sparse.py", line 162, in forward
    return F.embedding(
  File "/scratch/mlazos/pytorch/torch/nn/functional.py", line 2179, in embedding
    return handle_torch_function(
  File "/scratch/mlazos/pytorch/torch/overrides.py", line 1535, in handle_torch_function
    result = mode.__torch_function__(public_api, types, args, kwargs)
  File "/scratch/mlazos/pytorch/torch/_inductor/overrides.py", line 38, in __torch_function__
    return func(*args, **kwargs)
  File "/scratch/mlazos/pytorch/torch/nn/functional.py", line 2210, in embedding
    return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
  File "/scratch/mlazos/pytorch/torch/utils/_stats.py", line 20, in wrapper
    return fn(*args, **kwargs)
  File "/scratch/mlazos/pytorch/torch/_subclasses/fake_tensor.py", line 1060, in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
  File "/scratch/mlazos/pytorch/torch/_subclasses/fake_tensor.py", line 1139, in dispatch
    args, kwargs = self.validate_and_convert_non_fake_tensors(
  File "/scratch/mlazos/pytorch/torch/_subclasses/fake_tensor.py", line 1293, in validate_and_convert_non_fake_tensors
    return tree_map_only(
  File "/scratch/mlazos/pytorch/torch/utils/_pytree.py", line 266, in tree_map_only
    return tree_map(map_only(ty)(fn), pytree)
  File "/scratch/mlazos/pytorch/torch/utils/_pytree.py", line 196, in tree_map
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/scratch/mlazos/pytorch/torch/utils/_pytree.py", line 196, in <listcomp>
    return tree_unflatten([fn(i) for i in flat_args], spec)
  File "/scratch/mlazos/pytorch/torch/utils/_pytree.py", line 247, in inner
    return f(x)
  File "/scratch/mlazos/pytorch/torch/_subclasses/fake_tensor.py", line 1285, in validate
    raise Exception(
Exception: Please convert all Tensors to FakeTensors first or instantiate FakeTensorMode with 'allow_non_fake_inputs'. Found in aten.embedding.default(*(tensor([[-0.0102, -0.0615, -0.0265,  ..., -0.0199, -0.0372, -0.0098],
        [-0.0117, -0.0600, -0.0323,  ..., -0.0168, -0.0401, -0.0107],
        [-0.0198, -0.0627, -0.0326,  ..., -0.0165, -0.0420, -0.0032],
        ...,
        [-0.0218, -0.0556, -0.0135,  ..., -0.0043, -0.0151, -0.0249],
        [-0.0462, -0.0565, -0.0019,  ...,  0.0157, -0.0139, -0.0095],
        [ 0.0015, -0.0821, -0.0160,  ..., -0.0081, -0.0475,  0.0753]],
       device='cuda:0', grad_fn=<BroadcastBackward>), FakeTensor(FakeTensor(..., device='meta', size=(128, 128), dtype=torch.int64), cuda:0), 0), **{}) 

While executing %self_embeddings_word_embeddings : [#users=1] = call_module[target=self_embeddings_word_embeddings](args = (%input_ids,), kwargs = {})
Original traceback:
  File "/scratch/mlazos/transformers/src/transformers/models/bert/modeling_bert.py", line 1011, in <graph break in forward>
    embedding_output = self.embeddings(
  File "/scratch/mlazos/pytorch/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/scratch/mlazos/transformers/src/transformers/models/bert/modeling_bert.py", line 230, in forward
    inputs_embeds = self.word_embeddings(input_ids)
```

**Comment by mlazos** [comment]
Looks like I encountered: #95731

**Comment by bdhirsh** [comment]
I think this is a dupe of https://github.com/pytorch/pytorch/issues/93484 too (cc @yanboliang).

I tried repro'ing locally, and I could see that one of the example value inputs to aot autograd was an ordinary tensor, instead of a fake tensor. So dynamo is "missing" some inputs when it tries to wrap them in fake tensors.

**Comment by Lokiiiiii** [comment]
Fixed in PT 2.0 rc5
